### PR TITLE
Haewon - L102.ghg en usa s t y

### PR DESCRIPTION
**Pull request by d3y419**
#130

### CODE REVIEW

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 1]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
 
     # Load required inputs
     iso_GCAM_regID <- get_data(all_data, "common/iso_GCAM_regID")
-    IEA_flow_sector <- get_data(all_data, "energy/IEA_flow_sector")
-    IEA_product_fuel <- get_data(all_data, "energy/IEA_product_fuel")
+    IEA_flow_sector <- get_data(all_data, "energy/mappings/IEA_flow_sector")
+    IEA_product_fuel <- get_data(all_data, "energy/mappings/IEA_product_fuel")
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: 👍

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 2]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Merge-and-Match
-    # ===================================================
+    # Convert EPA GHG emissions inventory to Tg and aggregate by sector and fuel
+    EPA_FCCC_GHG_2005 %>% # start from EPA GHG
+      left_join(EPA_ghg_tech, by = "Source_Category") %>% # define category
+      select(-CO2) %>% # non-CO2 only
+      group_by(sector, fuel) %>%
+      summarize_if(is.numeric, sum, na.rm = FALSE) %>% # sum by sector and fuel
+      filter(!is.na(sector), !is.na(fuel)) %>% # delete NA sectors and fuel
+      mutate_all( funs( replace(., is.na(.), 1))) %>% # placeholder values for existing sectors that has no value.
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: No GCAM-style spacing next to parentheses and brackets please

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 3]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+      # NOTE: THIS IS A HACK. EPA DOESN'T HAVE EMISSIONS IN SOME SECTORS, WHEN EDGAR DOES. THIS WILL MAKE EMISSIONS PROPORTIONAL TO FUEL USE. NOT SURE IF THIS IS THE BEST STRATEGY.
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: Add this as an issue to follow up on later?

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 4]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+      L102.ghg_tg_USA_en_Sepa_F_2005 # GHG balance in 2005
+
+    if(OLD_DATA_SYSTEM_BEHAVIOR) {
+    # incorrect fuel name for transport from input mapping.
+    # organize energy balances in USA 2005
+        L101.in_EJ_R_en_Si_F_Yh %>% # start from energy balances
+        filter(GCAM_region_ID == gcam.USA_CODE) %>% # USA region only
+        select(-GCAM_region_ID) %>%
+        gather(variable, energy, -sector, -fuel, -technology) %>%
+        filter(variable == "X2005") %>%  # 2005 data only
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: Why is the data frame structured this way, i.e. why isn't this `year == 2005`?

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 5]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+                   by = c( "sector", "fuel", "technology" )) %>% # assign aggregate sector and fuel names
+        group_by(EPA_agg_sector, EPA_agg_fuel_ghg) %>%
+        summarize_if(is.numeric, sum) -> # sum by aggregate sector and fuel
+        L102.in_EJ_USA_en_Sepa_F_2005 # energy balance in 2005
+    }
+
+    # combine emissions and energy to get emission factors
+    L102.ghg_tg_USA_en_Sepa_F_2005 %>%
+      left_join(L102.in_EJ_USA_en_Sepa_F_2005, by = c( "sector" = "EPA_agg_sector", "fuel" = "EPA_agg_fuel_ghg" )) %>% # energy data and emission data joined
+      mutate(ch4_em_factor = CH4/energy) %>% # emission factor calculated
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: Here and below, spaces around `/` please

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 6]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
+      mutate(n2o_em_factor = N2O/energy) %>% # emission factor calculated
+      select(-CH4, -N2O, -energy) %>% # delete orignal data
+      arrange(fuel) %>%
+      mutate_all( funs( replace(., is.na(.), 0))) %>% # zero out NA
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: This is a pretty baroque way to do this--hard to see what's going on. Worth reshaping/changing for clarity? Also remove GCAM-style spacing

Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: Specifically, I think `mutate(value = if_else(is.na(value), 0, value))` (or something like that) would be clearer.

**Comment by d3y419** [comment]
@cahartin : suggested reviewer: Rachel

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 4 above)
Comment (cahartin) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: it looks like its because it's old_data_system_behavior?

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 6 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: done

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 5 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: done

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 4 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: it's for both old data behavior AND new behavior.
i inherited this df from: https://github.com/JGCRI/gcamdata/pull/256
which had Xyear and spreadform.
i could just mutate substr, but I end up throwing away this variable anyway.
how should i proceed?
how should i proceed?

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 3 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: will do.

same issue in https://github.com/JGCRI/gcamdata/pull/256 line 140

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 2 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: done

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 3 above)
Comment (d3y419) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: https://github.com/JGCRI/gcamdata/issues/327

**Comment by d3y419** [comment]
@cahartin @bpbond DONE!

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R (hunk 4 above)
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: Ugh. WTF. Well, I'm going to add this as an issue to follow up on, but for now we'll allow it.

File: R/zchunk_L102.ghg_en_USA_S_T_Y.R [hunk 7]
```diff
@@ -31,46 +31,77 @@ module_emissions_L102.ghg_en_USA_S_T_Y_DISABLED <- function(command, ...) {
...
-    # ===================================================
+    # Convert EPA GHG emissions inventory to Tg and aggregate by sector and fuel
+    EPA_FCCC_GHG_2005 %>% # start from EPA GHG
+      left_join(EPA_ghg_tech, by = "Source_Category") %>% # define category
+      select(-CO2) %>% # non-CO2 only
+      group_by(sector, fuel) %>%
+      summarize_if(is.numeric, sum, na.rm = FALSE) %>% # sum by sector and fuel
+      filter(!is.na(sector), !is.na(fuel)) %>% # delete NA sectors and fuel
+      mutate_all(funs(replace(., is.na(.), 1))) %>% # placeholder values for existing sectors that has no value.
+      # NOTE: THIS IS A HACK. EPA DOESN'T HAVE EMISSIONS IN SOME SECTORS, WHEN EDGAR DOES. THIS WILL MAKE EMISSIONS PROPORTIONAL TO FUEL USE. NOT SURE IF THIS IS THE BEST STRATEGY.
```
Comment (bpbond) [inline R/zchunk_L102.ghg_en_USA_S_T_Y.R]: @d3y419 @cahartin - add this as an issue to follow up later?

**Pull request merged**